android-exoplayer/src/main/java/com/imggaming/utils/BIFReader.java

```diff
@@ -0,0 +1,259 @@
+
+            sink = Okio.buffer(Okio.sink(destFile));
+            Buffer sinkBuffer = sink.buffer();
+
+            long totalBytesRead = 0;
+            int bufferSize = 8 * 1024;
+            for (long bytesRead; (bytesRead = source.read(sinkBuffer, bufferSize)) != -1; ) {
+                sink.emit();
+                totalBytesRead += bytesRead;
+            }
+
+        } finally {
+            if (response != null) {
+                response.close();
+            }
+            if (sink != null) {
+                sink.flush();
+                sink.close();
+            }
+            if (source != null) {
+                source.close();
+            }
+        }
+    }
+
+    // ToDO: Make it run on background thread
+    public Bitmap getImage(long time) {
+        Log.d(TAG, "getImage() time = " + time);
+
+        if (!initialised) {
+            Log.d(TAG, "getImage() not initialised");
+            return null;
+        }
+
+        if (file == null || !file.exists()) {
+            Log.e(TAG, "getImage() file error");
+            return null;
+        }
+
+        int offset = 0;
+
+        for (int i = 0; i < noOfImages; i++) {
+            long current = timeStamps.get(i) * multiplier;
+            if (current < time) {
+                offset = i;
+            } else {
+                break;
+            }
+        }
+
+        Bitmap bMap = cache.get(offset);
+        if (bMap != null) {
+            Log.d(TAG, "getImage() image in cache offset = " + offset);
+            return bMap;
+        }
+
+        Log.d(TAG, "getImage() image offset = " + offset);
+
+        bMap = getReadBitmap(offset);
+        return bMap;
+    }
+
+    private Bitmap getReadBitmap(int offset) {
+        FileInputStream fis = null;
+        Bitmap bMap = null;
+        int start = offsets.get(offset);
+        int end = offsets.get(offset + 1) - 1;
+
+        try {
+            fis = new FileInputStream(file);
+
+            byte[] data = new byte[end - start];
+
+            fis.skip(start);
+
+            fis.read(data);
+
+            bMap = BitmapFactory.decodeByteArray(data, 0, data.length);
+
+            cache.put(offset, bMap);
+
+            Log.d(TAG, "getReadBitmap() read offset = " + offset);
+
+        } catch (FileNotFoundException e) {
+            Log.e(TAG, "getReadBitmap() error ", e);
+        } catch (IOException e) {
+            Log.e(TAG, "getReadBitmap() error ", e);
+        } finally {
+            if (fis != null) {
+                try {
+                    fis.close();
+                } catch (IOException e) {
+                    e.printStackTrace();
+                }
+            }
+        }
+        return bMap;
+    }
+
+    public void close() {
+        cache.evictAll();
+        if (file != null && file.exists()) {
+            file.delete();
+        }
+    }
+
+    @Override
+    protected void finalize() throws Throwable {
+        close();
+        super.finalize();
```

Comment:
Our static analysis engine detects that your pull request includes a change of **explicit invocation of finalizer**. According to [Java doc](https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#finalize--), finalizer methods are designed **to be executed once, and only by the Virtual Machine (VM)**. If a connected set of objects beings finalizable, then the VM will invoke the finalize method on all the finalizable object, possibly at the same time in different threads. Thus, it is a particularly bad idea, in the finalize method for a class X, invoke finalize on objects referenced by X, because they may already be getting finalized in a separate thread.